Cater for using targeting in IPL

devenrao opened the pull request:
Following changes are done to pdbg repository

1) Do not use platform_attrs.H generated header file which uses pdbg_get/set attribute rather follow the hosboot way and use geneated fapi2platmacros.H which internally calls tryGetAttr and trySetAttr inside the macros.

2) Add missing attribute in rainier generated dts file ATTR_HW543384_WAR_MODE as this attribute is accessed in step 0 hwp.

3) Modify autogenerated [fapi2platattrmacros.H](https://github.com/devenrao/pdbg_refactor/pull/2/commits/f8ed09965dcbdda86ab58b69e38bd23f4fc3451f#diff-542e165fe7204e7e59bbf34db433b3b9d1d764213af35f21624730550e499d5c) to use constexpr and consteval over integer cast as the tryGetAttr, trySetAttr are template methods and we cannot use integer cast as part of compile time code generation for template methods.

4) Modify target service to loop through all the proc targets and set the HW_ACCESS_METHOD default pointer to SBEFIFO_ACCESS_PTR this way any invocation of HWP getcfam/putcfam getscom/putscom calls the sbefifo related methods. Direct access is used only when SBE is down.

devenrao committed:
target: endian conv for scalar types in DT access

Ensure scalar attribute values are correctly converted
from big-endian when reading data from the device tree.
This fixes potential issues with misinterpreted values
on little-endian platforms.

Change-Id: I2ecb9246cb365a9de4e63fa27ff67c17f25e03b8
Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>

devenrao committed:
target: standardize Google Test subproject wrap

Modified the Google Test wrap to follow the standard format used
across OpenBMC repositories such as sdbuplus. Ensures consistent
integration and usage of the testing framework within the project.

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I545d25b06706bd4ad9233012376ce63fd5907fbf

devenrao committed:
target: use local libfdt for local builds

- For local Meson builds, add libfdt as a subproject to provide the
  'fdt' dependency. This allows building targeting library without
  requiring system-installed dtc.

- For Yocto/Bitbake builds, rely on the dtc recipe and add DEPENDS
  on dtc, avoiding local subproject build and ensuring proper
  packaging.

- Avoid using dtc.wrap as it caused issues due to dtc being distributed
  as a tar.gz instead of a git repository.

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I2d4e13af157eae933d2bf093d54aea09f03d7502

devenrao committed:
target: move sample test app to test folder

- Relocate the sample application used to validate targeting changes
  from the main folder to the dedicated test folder.
- Update Meson build files accordingly to reflect the new location.

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I388ce32fda3b51f4dc4ac4138ca2fc7ed979a0d3

devenrao committed:
target: switch to rainier generated files

Test by switching to rainier auto-genrated files and
dtb file.

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: Ie48623bc697e01c3c9dd22a374deb98f04d7c7bd

devenrao committed:
target: add fapi to targeting attribute mapping

Add auto generated fapi to targeting attribute mapping
header file.

Change-Id: I52b7311f4ef0c4770933050b7ca3dc5eeb205077

devenrao committed:
target: add missing attributes used in HWP

ATTR_HW543384_WAR_MODE is used in clock test HWP and the
same is not generated so adding it manually

Change-Id: I1ea51bd2ab2481c317567404aa89451bc45553f8

devenrao committed:
target: cater for c-type array data types

ATTR_MC_PLL_BUCKET uses C-type array so catering
for the same in tryGetAttr and trySetAttr methods

Also removed private helper method setAttrImpl and using
trySetAttrImpl to avoid duplicate code.

Change-Id: I38a4db5103f43003376a284ddb4b885b5ca0c7cc

devenrao committed:
target: optimize get set attribute methods

optimized trySetAttrImpl tryGetAttrImpl methods which has
too many if..else blocks

Change-Id: I76192a8e2810a3f5f84bbe67828d7d7386db1491

devenrao committed:
target: fapi2 to targeting attribute mapping

- Add Fapi2ToTargetingAttr template to map fapi2::AttributeId to
  TARGETING::ATTRIBUTE_ID at compile-time.
- Define REGISTER_FAPI2_TO_TARGETING macro to simplify registration of
  each attribute mapping.
- Register mappings for SYS0, SYS1, PCI0, and PCI1 REFCLK receiver terms.
- Replaces previous enum-based approach, enabling type-safe, modern
  C++20 compliant attribute conversions.

Change-Id: I682c408de137d17119b5fce3928eac24580a3cd3

devenrao committed:
target: Set default access method for PROC target to SBEFIFO

Set the default hardware access method for all processor (PROC) targets
to SBEFIFO, which will be used for normal chip communication.

In scenarios where the BMC fails to communicate with the SBE (e.g.,
SBE not responding or FIFO timeout).

BMC will switch to the direct FSI access method to retrieve debug and
recovery data.

Change-Id: Ie9fabe4fe52bc857cf3fe15d05dcb566736b66cb

devenrao committed:
target: fix wrong namespace name used

Change-Id: Ifcfd2f35ea17e700d550bac573f7dfcb10e3428d

devenrao committed:
target: make seeprom select attributes writeable

Change-Id: I12faeafe3da66f3d037633f05c02fb10e8966203

devenrao committed:
target: removed repeating traces and fixed clang

Change-Id: I57092d51efdbe25ff2a599831f219ddde1d5aa8b

devenrao committed:
target: change default association type to childByPhysical

Previously, methods in Target used childByAffinity as the default
association type. This has been modified to childByPhysical to
maintain consistency with hostboot targeting and ensure uniform target
traversal behavior.

Change-Id: Ib831d643c039e729720100b8697738bcc5528f45

devenrao committed:
target: support endian conversion for enum-type attributes

Extended htobe_integral and betoh_integral utilities to handle enum types,
including enums defined with underlying types like uint8_t, uint16_t, etc.

This ensures attribute get/set operations correctly perform endian
conversion for both integral and enum-based attributes.

Tested
Test3: TYPE_PROC size
TYPE_PROC size 4

Test5: Endian conversion when read uint32_t value from dtb file
0x1

Change-Id: I5d9eb441a69bc0620b4f3033cbcc2f10cb402c98

devenrao committed:
target: fix invalid fd error in transport layer

Return FdHandle by value instead of returning pointer to a local
object, which was going out of scope and causing invalid file
descriptor errors during CFAM read/write operations.

Also added temporary debug logs for demo; these will be removed
in a later cleanup commit.

Change-Id: I1476472b351dbdbdc98d59e9dc84bf740b90df9e

devenrao committed:
target: cater for read/write hwas state struct

Special handling needs to be done for HwasState struct
when reading / writing to device tree

Test3: test hwas state size
Setting HWAS deconfiguredByEid 0x12345678
poweredOn: true, present: true, functional: true
Getting HWAS deconfiguredByEid 0x12345678
Test4: PredicatePostFoxExpr AttrVal and Attr

Change-Id: Ia3fb69d5b2404fc095d03c61aa2d4bf2d7ee2e06

deepakala-k commented on target/targeting/endian_utils.H:
```diff
@@ -0,0 +1,78 @@
+template <typename T> inline T betoh_integral(T v);
+
+// --- Unsigned specializations ---
+template <> inline uint16_t htobe_integral<uint16_t>(uint16_t v)
```
can we change the variable name to value. It has v and u in some methods. I thought u corresponds to unsigned value, but it also means value in this case

deepakala-k commented on target/targeting/endian_utils.H:
```diff
@@ -0,0 +1,78 @@
+{
+    return htobe64(v);
+}
+
```
Big endian to host begins here. Better to add a comment to separate them.

deepakala-k commented on target/targeting/xmltohb/fapi2platattrmacros.H:
```diff
@@ -0,0 +1,475 @@
+
+enum {
+
+ATTR_SYS0_REFCLOCK_RCVR_TERM =         TARGETING::ATTR_SYS0_REFCLOCK_RCVR_TERM,
```
We will replace this with the generated file

devenrao replied:
yes after this code is merged i will take latest files generated for device tree and will fix accordingly

deepakala-k commented on target/targeting/xmltohb/attributeenums.H:
```diff
@@ -145,6 +145,7 @@ enum ATTRIBUTE_ID
     ATTR_I2C_PARENT_PHYS_PATH                                      = 0xbad9735,
     ATTR_HW_ACCESS_METHOD                                          = 0xe4c8394,
     ATTR_HW_ACCESS_PTR                                             = 0xb34dd4a,
+    ATTR_HW543384_WAR_MODE                                         = 0xa450ad9,
```
workaround: need not review this change. It will be replaced by the auto generated file.

deepakala-k commented on target/targeting/xmltohb/fapi2platattrmacros.H:
```diff
@@ -2,6 +2,7 @@
 #ifndef FAPI2_FAPIPLATATTRMACROS_H
 #define FAPI2_FAPIPLATATTRMACROS_H
 #include <targeting/xmltohb/attributeenums.H>
+#include <attribute_ids.H>
```
Will be replace by autogen file. Need not review

deepakala-k commented on target/targetsvc/target_service.C:
```diff
@@ -34,6 +35,14 @@ void TargetService::init(const std::string& dtbPath)
+    PredicateAttrVal<ATTR_TYPE> pred(TYPE_PROC);
+    auto top = getTopLevelTarget();
+    for (auto&& tgt : getAssociated(top, AssociationType::childByPhysical,
+                                    RecursionLevel::all, &pred))
```
Some methods use RecursionLevel::all and some use immediate. Both works. But better to stick to one style. Hopefully, once we have the wrapper methods, we can avoid this?

devenrao replied:
if we want at same level then we need to use immediate and go through all levels we need to use all, all takes times so based on need need to use immediate or all

example
Node0
proc0
proc1
proc2

if we want to get procs under the node then better to use immediate

deepakala-k commented on target/targeting/xmltohb/attributetraits.H:
```diff
@@ -471,7 +471,7 @@ template<>
 {
     public:
-        enum { readable, notHbMutex, notFspMutex, fspAccessible };
+        enum { writeable, readable, notHbMutex, notFspMutex, fspAccessible };
```
workaround: to use the ATTR_BACKUP_SEEPROM_SELECT_WRITEABLE attribute

deepakala-k commented on target/targeting/target_tmpl.H:
```diff
@@ -282,7 +283,8 @@ bool Target::trySetAttrImpl(const T& attrValue)
+                           std::is_enum_v<TraitType>) &&
                           sizeof(TraitType) > 1)
             {
                 tmp = util::htobe_integral<TraitType>(tmp);
```
What if does not satisfy any constexpr? we need to capture and log or assert it? Is that condition possible?

devenrao replied:
good suggestion earlier hwasstate was escaping and going into this condition, now added statis assert so that it fails during compile time

deepakala-k commented on target/transport/transport.C:
```diff
@@ -228,46 +228,44 @@ int send_all(int fd, std::span<const std::byte> cmd)
+    auto fdOpt = prepareCfamAccess(target, addr);
+    if (!fdOpt)
     {
         return -1;
```
No traces?

deepakala-k commented on target/transport/transport.C:
```diff
@@ -228,46 +228,44 @@ int send_all(int fd, std::span<const std::byte> cmd)
+    auto fdOpt = prepareCfamAccess(target, addr);
+    if (!fdOpt)
     {
         return -1;
```
No traces? Not needed?

deepakala-k commented on target/targeting/common/entitypath.C:
```diff
@@ -127,17 +127,174 @@ std::string EntityPath::pathElementTypeAsString(TYPE type) const
             return "Sys";
         case TYPE_NODE:
             return "Node";
+        case TYPE_DIMM:
```
Wouldn't it be better to generate this using scripts? it will be difficule to maintain this way, right? and usually we have "sys" but here we have "Sys"  so can i add that method in xmltohb.pl script?

devenrao committed:
target: cater for reading std array into predicate

unit test case result
[ RUN      ] TargetServiceTest.TestPredicateAttrValProcFapiName
[       OK ] TargetServiceTest.TestPredicateAttrValProcFapiName (3 ms)

Change-Id: Ia635730d2d44007512709131a04b26e462415b7d

devenrao committed:
target: add performance analysis app

add perfornamce analysis app to check time it takes
to load and parse device tree

Change-Id: I8f61c103d54b4fbe9eb09c1c0ebdafc06712532f

dhruvibm committed:
Merge branch 'refactor-phal' into cbs
